### PR TITLE
DT-816: validate tokens

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/casenotes/config/TokenVerifyingAuthManager.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.security.oauth2.jwt.JwtDecoder
+import org.springframework.security.oauth2.server.resource.BearerTokenAuthenticationToken
+import org.springframework.security.oauth2.server.resource.InvalidBearerTokenException
+import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationProvider
+import org.springframework.stereotype.Component
+import org.springframework.web.reactive.function.client.WebClient
+
+@Suppress("SpringJavaInjectionPointsAutowiringInspection")
+@Component
+class TokenVerifyingAuthManager(jwtDecoder: JwtDecoder,
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/config/TokenVerifyingAuthManager.kt:18]: had to provide my own authentication manager as `JwtAuthenticationProvider` is final and can't be overridden

File: src/main/java/uk/gov/justice/hmpps/casenotes/config/TokenVerifyingAuthManager.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    if (tokenVerificationEnabled) {
+      val bearer = authentication as BearerTokenAuthenticationToken
+
+      // firstly check the jwt is still valid
+      val tokenDto = tokenVerificationApiWebClient.post().uri("/token/verify")
+          .header(HttpHeaders.AUTHORIZATION, "Bearer ${bearer.token}")
+          .retrieve()
+          .bodyToMono(TokenDto::class.java)
+          .blockOptional()
+      val tokenActive = tokenDto.map { it.active }.orElse(false)
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/config/TokenVerifyingAuthManager.kt:40]: have to check boolean value on return call, as status will always be 200

File: src/main/java/uk/gov/justice/hmpps/casenotes/config/TokenVerifyingAuthManager.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  override fun authenticate(authentication: Authentication): Authentication {
+
+    if (tokenVerificationEnabled) {
+      val bearer = authentication as BearerTokenAuthenticationToken
+
+      // firstly check the jwt is still valid
+      val tokenDto = tokenVerificationApiWebClient.post().uri("/token/verify")
+          .header(HttpHeaders.AUTHORIZATION, "Bearer ${bearer.token}")
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/config/TokenVerifyingAuthManager.kt:36]: have to manually add the header in here as is called before the `UserContextFilter` is setup

File: src/main/java/uk/gov/justice/hmpps/casenotes/config/WebClientConfiguration.kt [hunk 4]
```diff
@@ -23,34 +24,43 @@ import org.springframework.web.reactive.function.client.WebClient.Builder
...
-  fun oauthApiWebClient(builder: Builder): WebClient = createForwardAuthWebClient(builder, prop.oauthApiBaseUrl)
+  fun oauthApiWebClient(builder: Builder): WebClient = createForwardAuthWebClient(builder, oauthApiBaseUrl)
 
   @Bean
-  fun oauthApiHealthWebClient(builder: Builder): WebClient = createHealthClient(builder, prop.oauthApiBaseUrl)
+  fun oauthApiHealthWebClient(builder: Builder): WebClient = createHealthClient(builder, oauthApiBaseUrl)
 
   @Bean
-  fun tokenVerificationApiWebClient(builder: Builder): WebClient = createForwardAuthWebClient(builder, prop.tokenVerificationApiBaseUrl)
+  fun tokenVerificationApiWebClient(builder: Builder): WebClient = builder.baseUrl(tokenVerificationApiBaseUrl).build()
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/config/WebClientConfiguration.kt:52]: don't use forward header here as calls to the web client happen before the filter is configured

File: src/test/java/uk/gov/justice/hmpps/casenotes/controllers/ResourceTest.kt [hunk 5]
```diff
@@ -27,7 +28,7 @@ abstract class ResourceTest {
   @Autowired
   internal lateinit var jwtHelper: JwtAuthHelper
 
-  fun addBearerAuthorisation(user: String, roles: List<String>?): Consumer<HttpHeaders> {
+  fun addBearerAuthorisation(user: String, roles: List<String>? = listOf()): Consumer<HttpHeaders> {
```
Comment (mikehalmamoj) [inline src/test/java/uk/gov/justice/hmpps/casenotes/controllers/ResourceTest.kt:31]: No need for nullable any more?

Comment (petergphillips) [reply src/test/java/uk/gov/justice/hmpps/casenotes/controllers/ResourceTest.kt:31]: agreed, will fix in next PR.  Wanted to change one of the resource tests anyway to Kotlin

**Pull request merged**